### PR TITLE
Scaler fixes 4

### PR DESCRIPTION
**Pull request by daschuer**
This cleans up some code in linear scaler, a leftover from my fist aproach to include the vinylsoundemu there. 
It turns out, that it is best to stick with the vinylsoundemu in enginepregain and make it aware of scratching.

### CODE REVIEW

**Comment by daschuer** [comment]
@ywwg: Can you have a look at this?  Thank you.

**Comment by daschuer** [comment]
Introduced EnginePregain::setScratching()

**Comment by ywwg** [comment]
performance impact of one extra iteration over a buffer is unimportant.  The most maintainable solution is the most modular and readable solution.

**Comment by daschuer** [comment]
Please have a look at https://github.com/mixxxdj/mixxx/pull/552 the successor of this, which provides a complete ramping solution.

**Comment by ywwg** [comment]
Once https://github.com/mixxxdj/mixxx/pull/556 is in, we can write tests for all the situations you're working with (pause / unpause, pressing the reverse button, vinyl ramping, etc) and then we can see how this PR affects the output, if at all.

**Comment by daschuer** [comment]
OK https://github.com/mixxxdj/mixxx/pull/556 is merged now. Are you satisfied with the results? 
If yes, I would like to update the golden samples to the behavior introduced here and merge this.

**Comment by daschuer** [comment]
I would like to continue here. Is it OK for 1.12. This is a part of a series of fixes, in follow up PRs, fixing some real click issues as well.

**Comment by ywwg** [comment]
What clicks?  Can you write a test against current code that shows the clicks?  That's what it's for after all :)

**Comment by daschuer** [comment]
In this case it is not a click, it is a low frequency blob. You cannot here it with Laptop speakers. 

Thanks to your nice new tests tools, the test is already in place: 
scratchtest

You see the ramp of the new code to Zero when changing direction.
The original dark green curve looks not that bad, because the direction changes just at the sine wave maximum. You will have spikes if the sine wave is shifted by 90° 

You can see the issue better with real data: 

A scratched sample using Mixxx master, showing the issue. 
scatchmaster

The same scratched sample with this PR 
scratchmixxx

File: src/engine/enginepregain.cpp [hunk 1]
```diff
@@ -33,7 +33,9 @@ ControlObject* EnginePregain::s_pEnableReplayGain = NULL;
    A pregaincontrol is ... a pregain.
```
Comment (ywwg) [inline src/engine/enginepregain.cpp]: Remove this comment and put this text in the header file: "The pregain control alters the volume of the track based on several inputs, including user pregain adjustment, ReplayGain value, and vinyl-like adjustments in volume relative to playback speed."  I think that's all of the cases

File: src/engine/enginebufferscalelinear.cpp [hunk 2]
```diff
@@ -361,12 +352,17 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
...
                      frac * (static_cast<float>(ceil_sample[1]) -
                      static_cast<float>(floor_sample[1]));
+
         m_floorSampleOld[0] = floor_sample[0];
         m_floorSampleOld[1] = floor_sample[1];
 
+        // Smooth any changes in the playback rate over one buf_size
+        // samples. This prevents the change from being discontinuous and helps
+        // improve sound quality.
+        double rate_add = fabs((i * rate_diff / buf_size) + rate_old);
```
Comment (ywwg) [inline src/engine/enginebufferscalelinear.cpp]: const

File: src/engine/enginebufferscalelinear.cpp [hunk 3]
```diff
@@ -257,7 +254,7 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
     // Multiply by 2 because it is predicting mono rates, while we want a stereo
     // number of samples.
     // 0 is never the right answer
-    double unscaled_samples_needed = math_max<long>(2, unscaled_frames_needed * 2);
+    int unscaled_samples_needed = math_max<int>(2, unscaled_frames_needed * 2);
```
Comment (ywwg) [inline src/engine/enginebufferscalelinear.cpp]: const I think

File: src/engine/enginepregain.h [hunk 4]
```diff
@@ -31,10 +31,14 @@ class EnginePregain : public EngineObject {
     virtual ~EnginePregain();
 
     void setSpeed(double speed);
+    void setScratching(bool scratching);
```
Comment (ywwg) [inline src/engine/enginepregain.h]: documentation:
// If the user is scratching and the record reverses direction, the volume will be ramped to zero and back up again to mimic a vinyl scratch.  If the user is not scratching and the direction is reversed (e.g. reverse button is pressed), the audio will be immediately reversed without a ramp to zero.

File: src/engine/enginebuffer.cpp [hunk 5]
```diff
@@ -479,16 +479,19 @@ void EngineBuffer::setNewPlaypos(double newpos) {
...
-{
+double EngineBuffer::getSpeed() {
     return m_speed_old;
 }
 
+bool EngineBuffer::getScratching() {
+    return m_scratching_old;
+}
+
+
```
Comment (ywwg) [inline src/engine/enginebuffer.cpp]: extra newline here

File: src/engine/enginebufferscalelinear.cpp (hunk 3 above)
Comment (daschuer) [inline src/engine/enginebufferscalelinear.cpp]: no, is is reassigned below.

**Comment by daschuer** [comment]
Notes addressed and tests are running. Ready to merge?

**Comment by daschuer** [comment]
If there are no further objection, I will merge this soon.

**Comment by ywwg** [comment]
LGTM

**Comment by daschuer** [comment]
Thank you for review.

File: src/engine/enginebufferscalelinear.cpp [hunk 6]
```diff
@@ -238,12 +220,11 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
...
     double frames = 0;
-
     // We're calculating frames = 2 samples, so divide remaining buffer by 2;
-    for (int j = 0; j < iRateLerpLength; j += 2) {
-        frames += fabs((rate_add_diff * static_cast<float>(j)) /
-                        static_cast<float>(iRateLerpLength) + rate_add_old);
+    for (int j = 0; j < buf_size / 2; ++j) {
+        frames += (j * 2 * rate_diff / buf_size) + rate_old;
     }
+    frames = abs(frames);
```
Comment (rryan) [inline src/engine/enginebufferscalelinear.cpp]: Hey @daschuer, @ywwg -- this uses abs(int) so it loses the decimal part of frames. Was the double -> int conversion intentional?

Comment (ywwg) [inline src/engine/enginebufferscalelinear.cpp]: probably not.  we're supposed to take the floor, so there's a 50/50 chance we end up with an extra frame

Comment (rryan) [inline src/engine/enginebufferscalelinear.cpp]: Any thoughts @daschuer ?

Comment (daschuer) [inline src/engine/enginebufferscalelinear.cpp]: This is a bug it should be fabs()

Comment (daschuer) [inline src/engine/enginebufferscalelinear.cpp]: fixed in 82e43d9

**Pull request merged**